### PR TITLE
EcsRunLauncher - propagate ECS task definition tags to running tasks

### PR DESCRIPTION
**Pull request by dixel**
Hi dear Dagster maintainers!

## Summary

This change propagate tags from ECS TaskDefinition to running tasks. This allows better cost tracking and makes task definition tags more usable ([aws doc](https://docs.aws.amazon.com/AmazonECS/latest/APIReference/API_RunTask.html#ECS-RunTask-request-propagateTags), [boto3 doc](https://boto3.amazonaws.com/v1/documentation/api/latest/reference/services/ecs.html#ECS.Client.run_task)).
Creating it as a draft for now, since I still would like to test this somehow. But keen to hear initial feedback if this change actually makes sense.

## Test Plan

I'll try to check existing tests and see if it's possible to check if tasks spinned up by EcsRunLauncher inherit the tags from task definitions. Removing WIP when it's done.

## Checklist

- [x] My change requires a change to the documentation and I have updated the documentation accordingly.
- [x] I have added tests to cover my changes.

### CODE REVIEW

File: python_modules/libraries/dagster-aws/dagster_aws/ecs/launcher.py [hunk 1]
```diff
@@ -118,6 +118,7 @@ def launch_run(self, context: LaunchRunContext) -> None:
                 }
             },
             launchType="FARGATE",
+            propagateTags="TASK_DEFINITION",
```
Comment (jmsanders) [inline python_modules/libraries/dagster-aws/dagster_aws/ecs/launcher.py:121]: I'm curious how this behaves when https://aws.amazon.com/blogs/compute/migrating-your-amazon-ecs-deployment-to-the-new-arn-and-resource-id-format-2/ isn't enabled yet. We recently had to add some changes to stay backwards compatible with the older ARN format which doesn't support tags:

https://github.com/dagster-io/dagster/pull/5651

So we might want to play it safe and only conditionally add this param if we know for sure the account uses the newer long ARN format.

**Review by jmsanders — CHANGES_REQUESTED** [review]
Thanks for the contribution! The existing tests run against a [fake version of ECS](https://github.com/dagster-io/dagster/blob/master/python_modules/libraries/dagster-aws/dagster_aws_tests/ecs_tests/stubbed_ecs.py) and I suspect for your purposes, the stubs might already work without any changes. Adding an assertion here might be sufficient:

https://github.com/dagster-io/dagster/blob/a5ca15ed7f63c04c967da007b520421d59353fca/python_modules/libraries/dagster-aws/dagster_aws_tests/ecs_tests/launcher_tests/test_launching.py#L35-L41

File: python_modules/libraries/dagster-aws/dagster_aws/ecs/launcher.py (hunk 1 above)
Comment (dixel) [reply python_modules/libraries/dagster-aws/dagster_aws/ecs/launcher.py]: Thanks for pointing this out. I can't really reproduce this behaviour, but I assume it would fail (https://github.com/cloudposse/terraform-aws-ecs-alb-service-task/issues/39). So far incorporated it similarly to the way it is done in `_set_run_tags` method and updated the stub to emulate AWS API behaviour.

**Comment by dixel** [comment]
> Thanks for the contribution! The existing tests run against a [fake version of ECS](https://github.com/dagster-io/dagster/blob/master/python_modules/libraries/dagster-aws/dagster_aws_tests/ecs_tests/stubbed_ecs.py) and I suspect for your purposes, the stubs might already work without any changes. Adding an assertion here might be sufficient:
> 
> https://github.com/dagster-io/dagster/blob/a5ca15ed7f63c04c967da007b520421d59353fca/python_modules/libraries/dagster-aws/dagster_aws_tests/ecs_tests/launcher_tests/test_launching.py#L35-L41

@jmsanders Thanks for your review!
*TL;DR*
The code added to StubbedEcs validates whether adding this parameter is safe for both old/new ARN's. On the test level I couldn't find any visible artifacts that could be tested after applying `propagateTags`.

I tried a few options to test the expected behaviour (tags from task_definition being propagated to a task).
First I tried to set tags on task_definition level, but ECS stub doesn't support this. I tried passing `tags: ...` to `register_task_definition` method and also tried using `ecs.tag_resource` for task definition.
After that I tried to simply validate if this parameter gets propagated to the task level. Unfortunately AWS API doesn't leave any traces of this parameter after task creation. Also thought about getting around this by patching `StubbedEcs` but didn't find any non-hacky way of doing that without breaking the API/code consistency and isolation.
Maybe you have some ideas.

File: python_modules/libraries/dagster-aws/dagster_aws/ecs/launcher.py [hunk 2]
```diff
@@ -107,19 +107,30 @@ def launch_run(self, context: LaunchRunContext) -> None:
...
             },
-            launchType="FARGATE",
-        )
+            "launchType": "FARGATE",
+        }
+
+        try:
+            response = self.ecs.run_task(
+                **ecs_run_task_args,
+                propagateTags="TASK_DEFINITION",
```
Comment (jmsanders) [inline python_modules/libraries/dagster-aws/dagster_aws/ecs/launcher.py:129]: Instead of always propagating tags from the task definition, I think we should make this configurable in `dagster.yaml`. Similar to how we added the ability to configure secrets recently: https://github.com/dagster-io/dagster/pull/5614/commits/c1fe46c29913ce38a17d20c141217572136e375c

That way, users can choose whether to set this to `TASK_DEFINITION` or leave it blank (the default). Perhaps the config should actually be a boolean - if `propogate_tags == True`, we'll include `propagateTags="TASK_DEFINITION"`; otherwise we won't.

File: python_modules/libraries/dagster-aws/dagster_aws_tests/ecs_tests/test_stubbed_ecs.py [hunk 3]
```diff
@@ -272,6 +275,14 @@ def test_run_task(ecs, subnet):
     )
     assert response["tasks"][0]["overrides"]["containerOverrides"][0]["command"] == ["ls"]
 
+    # fails if old ARN schema is used and `propagateTags` parameter is set
+    if task_long_arn_format == "disabled":
+        with pytest.raises(ClientError):
+            ecs.run_task(taskDefinition="container", propagateTags="TASK_DEFINITION")
+    else:
+        with pytest.raises(ClientError):
+            ecs.run_task(taskDefinition="container", propagateTags="WRONG_PARAMETER")
```
Comment (jmsanders) [inline python_modules/libraries/dagster-aws/dagster_aws_tests/ecs_tests/test_stubbed_ecs.py:284]: This does a good job of testing the various failure scenarios, but I don't think it adequately tests the happy path. I'd suggest something akin to:

```py
    # add some tags to a task definition - you might need to stub out some more
    # of `register_task_definition()` to make this happen

    # tags aren't propagated if the old ARN format is being used
    if task_long_arn_format == "disabled":
        with pytest.raises(ClientError):
            ecs.run_task(taskDefinition="container", propagateTags="TASK_DEFINITION")
            
    # tags aren't propagated if we pass an invalid value
    with pytest.raises(ClientError):
        ecs.run_task(taskDefinition="container", propagateTags="SERVICE")
    
    response = ecs.run_task(taskDefinition="container")
    assert not response["task"][0].get("tags")
    
    response = ecs.run_task(taskDefinition="container", propagateTags="TASK_DEFINITION")
    tags = response["tasks"][0].get("tags")
    # and then assert the tags match what you tagged the task definition with
```

File: python_modules/libraries/dagster-aws/dagster_aws/ecs/launcher.py [hunk 4]
```diff
@@ -107,19 +107,30 @@ def launch_run(self, context: LaunchRunContext) -> None:
...
                     "assignPublicIp": "ENABLED",
                     "securityGroups": metadata.security_groups,
                 }
             },
-            launchType="FARGATE",
-        )
+            "launchType": "FARGATE",
+        }
+
+        try:
```
Comment (jmsanders) [inline python_modules/libraries/dagster-aws/dagster_aws/ecs/launcher.py:126]: Can we also extend this test:

https://github.com/dagster-io/dagster/blob/c1292165f1daf7b747ffae60013188b69171eae2/python_modules/libraries/dagster-aws/dagster_aws_tests/ecs_tests/launcher_tests/test_launching.py#L86

To show that not only our stub mimics ECS correctly, but that we're interacting with our stub correctly? I'm thinking we add a few tags to our task definition, we configure our launcher to propagate tags, we launch a run, and we check that our tags from the task definition were copied over.

**Review by jmsanders — CHANGES_REQUESTED** [review]
A few suggestions on how to make these tests a bit more robust. The only behavioral change is that I think we should make this configurable.

Feel free to ping me if you want to pair through any of this at all. The ECS stubs we use can be a bit complicated.

**Comment by cmpadden** [comment]
Closing due to inactivity; please re-open if necessary, thank you
